### PR TITLE
compiler: Represent pruned scopes instead of inlining

### PR DESCRIPTION
**Pull request by josephsavona**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #29820
* #29810
* #29790
* #29789
* __->__ #29781

There are a few places where we want to check whether a value actually got memoized, and we currently have to infer this based on values that "should" have a scope and whether a corresponding scope actually exists. This PR adds a new ReactiveStatement variant to model a reactive scope block that was pruned for some reason, and updates all the passes that prune scopes to instead produce this new variant.

### CODE REVIEW

File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/BuildReactiveBlocks.ts [hunk 1]
```diff
@@ -183,6 +183,7 @@ function visitBlock(context: Context, block: ReactiveBlock): void {
         context.append(stmt, stmt.label);
         break;
       }
+      case "pruned-scope":
```
Comment (victor-0x29a) [inline compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/BuildReactiveBlocks.ts:186]:
```suggestion
```

Unecessary case

**Review by victor-0x29a — COMMENTED** [review]
some adjust

File: compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/BuildReactiveBlocks.ts (hunk 1 above)
Comment (josephsavona) [reply compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/BuildReactiveBlocks.ts:186]: No, this is necessary. We use exhaustive switch cases with Typescript to make sure we don’t silently fail to handle cases.

**Review by gsathya — APPROVED** [review]
yeah this makes sense

**Pull request merged**